### PR TITLE
Distributed Transaction: Action on commit prepared or redo prepared failure

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR make changes to transaction recovery log state in resource managers.

On a commit prepared failure it checks if the error is retryable or not. 
For a non-retryable error, the state is update to `FAILED`, error metric is updated and then they are are not prepared later. 

Same check happens on redo of prepared logs on a failure.

Once, the state is updated to `Failed`, a user intervention is needed to evaluate the root cause of the issue.
This is a highly unlikely scenario and could have happened due to a bug in the system.

## Related Issue(s)

- #16245 

## Checklist

-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletmanager/rpc_vreplication_test.go [hunk 1]
```diff
@@ -302,6 +302,7 @@ func TestCreateVReplicationWorkflow(t *testing.T) {
 // results returned. Followed by ensuring that SwitchTraffic
 // and ReverseTraffic also work as expected.
 func TestMoveTablesUnsharded(t *testing.T) {
+	t.Skip("Skipping test temporarily as it is flaky on CI, pending investigation")
```
Comment (systay) [inline go/vt/vttablet/tabletmanager/rpc_vreplication_test.go:305]: don't we just want to make the workflow not required?

File: go/vt/vttablet/tabletserver/tx_engine.go [hunk 2]
```diff
@@ -411,58 +411,129 @@ func (te *TxEngine) shutdownLocked() {
...
 		}
-		if txid > maxid {
-			maxid = txid
+	}
+
+outer:
+	for _, preparedTx := range prepared {
+		var conn *StatefulConnection
+
+		txID, _ := dtids.TransactionID(preparedTx.Dtid)
```
Comment (systay) [inline go/vt/vttablet/tabletserver/tx_engine.go:442]: I don't get it - why are we ignoring the error here?

Comment (systay) [reply go/vt/vttablet/tabletserver/tx_engine.go:442]: After talking it over with H - maybe we could solve this by extracting the inside of the loop into a method, and using a defer in that method to check errors

File: go/vt/vttablet/tabletserver/tx_engine_test.go [hunk 3]
```diff
@@ -603,3 +606,75 @@ func TestTxEngineFailReserve(t *testing.T) {
...
+	cfg.DB = newDBConfigs(db)
+	env := tabletenv.NewEnv(vtenv.NewTestEnv(), cfg, "TabletServerTest")
+	env.Config().TwoPCEnable = true
+	env.Config().TwoPCAbandonAge = 5
+	te := NewTxEngine(env, nil)
+	te.AcceptReadWrite()
+
+	tcases := []struct {
+		receivedErr  error
+		nonRetryable bool
```
Comment (systay) [inline go/vt/vttablet/tabletserver/tx_engine_test.go:623]: nitpick: I think it's easier to understand if this was `retryable` instead

**Review by systay — APPROVED** [review]

File: go/vt/vttablet/tabletmanager/rpc_vreplication_test.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vttablet/tabletmanager/rpc_vreplication_test.go:305]: I just took the commit from rohit's PR. 
I do not need this change anymore as it is already in `main` now.

Making CI not required make other tests also being ignored.

**Review by GuptaManan100 — APPROVED** [review]
LGTM

**Pull request merged**